File: tp_tda/entregables/9/main.py
import heapq
import sys

def main():
    def cost(a, b):
        return sum(min((int(a[i]) - int(b[i])) % 10, (int(b[i]) - int(a[i])) % 10) for i in range(4)) # Calculo el costo entre dos llaves
    
    def agmin_prim(start_vertex): # Funcion Prim que me va sumando el peso de todo el AGM
            visited = {v: False for v in keys}
            min_heap = []
            agm = []
            agm_weight = 0
            
            def add_edges(node):
                visited[node] = True
                for neighbor in keys:
                    if neighbor != node:
                        if not visited[neighbor]:
                            edge = (cost(node, neighbor), node, neighbor)
                            heapq.heappush(min_heap, edge)
            
            add_edges(start_vertex)
            
            while min_heap:
                weight, node1, node2 = heapq.heappop(min_heap)
                if not visited[node2]:
                    visited[node2] = True
                    agm.append(((node1, node2), weight))
                    agm_weight += weight
                    add_edges(node2)
            
            return agm_weight
    
    #Comienzo a guardar el input
    cases = int(input())
    i=0
    test_cases = []
    for line in sys.stdin:
        if i == cases:
            break
        test=line.split()
        num_keys = int(test.pop(0))
        keys = test
        test_cases.append((num_keys, keys))
        i+=1
    # Termine de guardar el input
    
    rolls = []
    for num_keys, keys in test_cases: 
        
        zero_cost_list = [((0000, vertex), cost("0000", vertex)) for vertex in keys] # Calculo el costo de todos los nodos al 0000
        min_edge, min_weight = min(zero_cost_list, key=lambda x: x[1]) # Calcúlo cual es la llave mas cercana al 0000
            
        agm_weight = agmin_prim(min_edge[1]) # Hago el Prim desde el nodo mas cercano al 0000
        rolls.append(agm_weight+min_weight) # Guardo el peso del AGM y le sumo el peso al 0000
    
    # Imprimo los resultados
    for r in rolls:
        print(r)

main()

File: tp_tda/entregables/9/main.cc
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>
#include <algorithm>

using namespace std;

int cost(string a, string b) {
    int totalCost = 0;
    for (int i = 0; i < 4; ++i) {
        int digitA = a[i] - '0';
        int digitB = b[i] - '0';
        totalCost += min(abs(digitA - digitB), 10 - abs(digitA - digitB));
    }
    return totalCost;
}

int agmin_prim(string start_vertex, unordered_map<string, bool>& visited, vector<string>& keys) {
    priority_queue<pair<int, pair<string, string>>, vector<pair<int, pair<string, string>>>, greater<pair<int, pair<string, string>>>> min_heap;
    vector<pair<string, string>> agm;
    int agm_weight = 0;

    auto add_edges = [&](string node) {
        visited[node] = true;
        for (const auto& neighbor : keys) {
            if (neighbor != node && !visited[neighbor]) {
                int edgeCost = cost(node, neighbor);
                min_heap.push({edgeCost, {node, neighbor}});
            }
        }
    };

    add_edges(start_vertex);

    while (!min_heap.empty()) {
        auto top = min_heap.top();
        min_heap.pop();
        auto node1 = top.second.first;
        auto node2 = top.second.second;
        if (!visited[node2]) {
            visited[node2] = true;
            agm.push_back({node1, node2});
            agm_weight += top.first;
            add_edges(node2);
        }
    }

    return agm_weight;
}

int main() {
    
    ios_base::sync_with_stdio(0);
	cin.tie(0);
    
    int cases;
    cin >> cases;

    vector<pair<int, vector<string>>> test_cases;
    for (int i = 0; i < cases; ++i) {
        int num_keys;
        cin >> num_keys;
        vector<string> keys(num_keys);
        for (int j = 0; j < num_keys; ++j) {
            cin >> keys[j];
        }
        test_cases.push_back({num_keys, keys});
    }

    vector<int> rolls;
    for (const auto& [num_keys, keys] : test_cases) {
        vector<pair<string, int>> zero_cost_list;
        for (const auto& vertex : keys) {
            zero_cost_list.push_back({vertex, cost("0000", vertex)});
        }
        auto min_edge = min_element(zero_cost_list.begin(), zero_cost_list.end(), [](const pair<string, int>& a, const pair<string, int>& b) {
            return a.second < b.second;
        });
        int min_weight = min_edge->second;

        unordered_map<string, bool> visited;
        vector<string> keys_copy = keys;
        int agm_weight = agmin_prim(min_edge->first, visited, keys_copy);
        rolls.push_back(agm_weight + min_weight);
    }

    for (int r : rolls) {
        cout << r << endl;
    }

    return 0;
}